### PR TITLE
Run mazerunner scenarios on CI

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

Run mazerunner scenarios on CI.

## Changeset

- Adds mazerunner scenarios to Travis CI environment matrix, running on API 21 emulator
- Alter scenarios to relaunch the app rather than start the activity manually. This adds a pre-requisite wait of X seconds that if not present would cause scenarios to fail.
- Add wait time of 15s after launching activity when running on Travis CI, due to slowness of emulator
- Reduce specificity of StrictMode scenarios (format of exception message differs across API levels)
- Make session scenarios more deterministic by flushing sessions (`SessionSetAutoCaptureScenario` tests that a session is delivered naturally without these test hooks)

## Tests

Ran tests on CI.

### Alternative Approaches

Running suite locally, or only on specific builds, rather than on every commit.

### Outstanding Questions

Running mazerunner scenarios on Travis-CI is really slow as we are limited to ARM emulators. I'm quite concerned that in the long-term this approach will prevent us from running against different API targets, and in the short-term may lead to further congestion of our build queue.

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
 source "https://rubygems.org"
 
-gem 'bugsnag-maze-runner', git: 'git@github.com:bugsnag/maze-runner'
+gem 'bugsnag-maze-runner', :git => 'https://github.com/bugsnag/maze-runner'
```
Comment (fractalwrench) [inline Gemfile]: allows Travis to clone the repo

File: features/strict_mode.feature [hunk 2]
```diff
@@ -5,7 +5,6 @@ Scenario: StrictMode DiscWrite violation
     Then I should receive a request
     And the request is a valid for the error reporting API
     And the exception "errorClass" equals "android.os.StrictMode$StrictModeViolation"
-    And the exception "message" equals "policy=262145 violation=1"
```
Comment (fractalwrench) [inline features/strict_mode.feature]: The policy value differs on the Travis CI emulator as it's a different API level, and all we really care about here is that the metadata value was set

File: mazerunner/src/main/java/com/bugsnag/android/TestHarnessHooks.kt [hunk 3]
```diff
@@ -2,12 +2,18 @@ package com.bugsnag.android
...
  * Accesses the session tracker and flushes all stored sessions
  */
-internal fun flushAllSessions() {
-    Bugsnag.getClient().sessionTracker.flushStoredSessions()
+internal fun flushAllSessions(client: Client) {
+    val thread = HandlerThread("HandlerThread")
+    thread.start()
+    Handler(thread.looper).post {
+        client.sessionTracker.flushStoredSessions()
+    }
```
Comment (fractalwrench) [inline mazerunner/src/main/java/com/bugsnag/android/TestHarnessHooks.kt]: Avoids performing network operations on the main thread, that could have caused scenarios to fail intermittently

File: mazerunner/src/main/java/com/bugsnag/android/mazerunner/scenarios/CustomClientSessionFlushScenario.kt [hunk 4]
```diff
@@ -14,14 +14,16 @@ internal class CustomClientSessionFlushScenario(config: Configuration,
                                                 context: Context) : Scenario(config, context) {
 
     override fun run() {
```
Comment (fractalwrench) [inline mazerunner/src/main/java/com/bugsnag/android/mazerunner/scenarios/CustomClientSessionFlushScenario.kt]: This scenario now only creates one session, as it is supposed to test that a manual session is cached on disk then delivered.

File: mazerunner/src/main/java/com/bugsnag/android/TestHarnessHooks.kt (hunk 3 above)
Comment (kattrali) [reply mazerunner/src/main/java/com/bugsnag/android/TestHarnessHooks.kt]: I don't entirely understand what this function is for; aren't sessions delivered automatically?

**Comment by fractalwrench** [comment]
Closing as superseded by work in #369

**Pull request closed without merging**